Allow stoa to call getPreimage and improve TestAgora

Geod24 opened the pull request:
Pretty straightforward, see each commit for details.

Geod24 committed:
Implement getPreimage endpoint in the test Agora

Geod24 committed:
TestAgora: Simplify /block_height endpoint

Reduce the number of express method calls we need to do,
and also uses 'Height' directly which is closer to the real code.

Geod24 committed:
TestAgora: Return a Bad Request on missing or invalid query param

Returning 200 doesn't quite mirror or Agora currently behave (and should behave).

Geod24 committed:
TestAgora: More parameter validation for GET /block_height

The code was previously silently replacing invalid values with valid ones.
This usually hides bug in user code and is not a desired outcome.

TrustHenry commented on src/modules/agora/AgoraClient.ts:
```diff
@@ -108,6 +108,32 @@ export class AgoraClient implements FullNodeAPI
+    /**
+     * Request the latest PreImageInfo for an enrollment key
+     *
+     * @param {enroll_key} Hash of the UTXO used by the validator
```
```suggestion
     * @param enroll_key{Hash} Hash of the UTXO used by the validator
```

TrustHenry replied:
Please add @returns document.

Geod24 replied:
Looking at the official documentation, it should be `@param {type} name`

https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#param-and-returns

Geod24 replied:
Fixed documentation

TrustHenry commented on tests/Utils.ts:
```diff
@@ -75,32 +75,52 @@ export class TestAgora
+
+            if (this.blocks.length < 1)
+            {
+                res.status(500).json({ errorMessage: 'No data on server' });
```
`errorMessage:`
Now should we apply all response error messages key in common?

Geod24 replied:
No, I should have used `statusText`

MichaelKim20 replied:
A `statusMessage` is used in Agora.
https://github.com/bpfkorea/vibe.d/blob/master/web/vibe/web/rest.d#L1466

TrustHenry commented on src/modules/agora/AgoraClient.ts:
```diff
@@ -108,6 +108,33 @@ export class AgoraClient implements FullNodeAPI
+     * @param enroll_key Hash of the UTXO used by the validator
+     * @returns A `PreImageInfo` matching this `enroll_key`
+     */
+    public getPreimage (enroll_key: Hash): Promise<PreImageInfo>
```
Do you want to change the method of collecting pre-image data from `put` to `get` request?

TrustHenry replied:
Or is it just for testing?

Geod24 replied:
No, we keep Agora's push approach, this is here for a fallback.

TrustHenry reviewed (approved):
LGTM

MichaelKim20 commented on src/modules/agora/AgoraClient.ts:
```diff
@@ -83,9 +83,9 @@ export class AgoraClient implements FullNodeAPI
      * @param max_blocks - The maximum number of block to request
      */
-    public getBlocksFrom (block_height: Height, max_blocks: number): Promise<Array<any>>
+    public getBlocksFrom (block_height: Height, max_blocks: number): Promise<Block[]>
```
Does it automatically convert to an instance of Block?

MichaelKim20 reviewed (commented):
@Geod24 : That's a great suggestion. I requested some revisions. Please check.

Geod24 commented:
Comments fixed:
- The first commit was lacking and has been moved to #245 
- The second commit missed a `reviver` call, added it;
- Changed `errorMessage` to the correct `statusText`;

Geod24 commented:
Anything else ?

MichaelKim20 reviewed (approved):
LGTM
